### ISSUE TITLE
Startup issues with ijavascript

### ISSUE THREAD

**Issue by lgeiger** [nteract/hydrogen#591]
**Description:**

If I want to use IJavascript in Hydrogen almost everytime the kernel won't properly start and only display the spinner. After restarting the IJavascript kernel everything works as expected.

This may be related to #584, #468 and #88 but for me it only happens with IJavascript.

I guess this could be caused by a race condition where the sockets are not ready to receive the execute request.

If I change our [startup script](https://github.com/nteract/hydrogen/blob/master/lib/kernel-manager.js#L131-L144) to
```javascript
launchSpec(kernelSpec, spawnOptions)
    .then(({ config, connectionFile, spawn }) => {
      const kernel = new ZMQKernel(
        kernelSpec, grammar,
        config, connectionFile,
        spawn,
      );

      setTimeout(() => {
        this.setRunningKernelFor(grammar, kernel);

        this._executeStartupCode(kernel);

        if (onStarted) onStarted(kernel);
      }, 1000);
    });
```
the issue disappears.
Is it necessary to wait for the [sockets to be connected](https://github.com/nteract/hydrogen/blob/master/lib/zmq-kernel.js#L95-L98)?

@n-riesco @rgbkrk Any idea what else could cause this issue?

**Steps to Reproduce:**

1. Start fresh Atom instance
2. Run javascript file with IJavascript -> only see the spinner
3. Restart IJavascript kernel via statusbar -> everything works fine

<details>
<summary><strong>Logs and Versions:</strong></summary>
A failed kernel startup:
<img width="619" alt="bildschirmfoto 2017-01-21 um 23 49 32" src="https://cloud.githubusercontent.com/assets/13285808/22178361/58c8e862-e034-11e6-89cd-40cd14dc683f.png">

A successful kernel startup:
<img width="619" alt="bildschirmfoto 2017-01-21 um 23 51 36" src="https://cloud.githubusercontent.com/assets/13285808/22178371/a1ab795a-e034-11e6-9a2c-cf0749c6a87e.png">
<img width="619" alt="bildschirmfoto 2017-01-21 um 23 51 41" src="https://cloud.githubusercontent.com/assets/13285808/22178372/a1ac1928-e034-11e6-858e-38a4a9f3ee4e.png">

**Versions:**

 macOS 10.12.2
```bash
$ atom --version
Atom    : 1.14.0-beta2
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
```
</details>

**Comment by rgbkrk** [comment]
Whoa, seems like we should definitely handle the callback for when it's really connected - especially before we try to send any initial messages.

**Comment by n-riesco** [comment]
OMG! It looks like #88 is back (I really was hoping it'd go away with prebuilt zeromq).

Unfortunately, #88 is very hard to reproduce. I've just tried and I can't reproduce in a machine with:

```
$ apm --version
apm  1.12.9
npm  3.10.5
node 4.4.5
python 2.7.12
git 2.7.4

$ ijs --version
5.0.17

$ apm list | grep Hydrogen
├── Hydrogen@1.5.3
```

If I remember correctly from my experiments with #88, when #88 happens, the `connect` event never gets sent on the IO socket.

@lgeiger Does your fix work when you replace `setTimeout` with `nextTick`? If it doesn't, what's the minimum wait that works?

**Comment by n-riesco** [comment]
@lgeiger I missed the log messages in your first post. There is an unexpected stderr message in the IJavascript launcher.

Since I can't reproduce this issue. Please, could check whether replacing `log("startupCallback:", this.startupCallback);` [here](https://github.com/n-riesco/ijavascript/blob/master/lib/kernel.js#L88) with `log("startupCallback:", arguments.callee);` fixes the issue for you?

**Comment by lgeiger** [comment]
Thanks for taking a look at this

> Unfortunately, #88 is very hard to reproduce.

It happens for me approx. 80% of the time. A simple kernel restart fixes it but it's still annoying 😆 

> If I remember correctly from my experiments with #88, when #88 happens, the connect event never gets sent on the IO socket.

Based on the logs, the IO socket is connected (or at least emits a `connected` event)

> Does your fix work when you replace setTimeout with nextTick? If it doesn't, what's the minimum wait that works?

Unfortunately not. Estimating the wait is also a bit tricky since the issue doesn't reproduce all the time. But 1s is definitely overkill.

> replacing log("startupCallback:", this.startupCallback); here with log("startupCallback:", arguments.callee); fixes the issue for you?

Unfortunately this doesn't help either.

**Comment by n-riesco** [comment]
> > replacing log("startupCallback:", this.startupCallback); here with log("startupCallback:", arguments.callee); fixes the issue for you?

> Unfortunately this doesn't help either.

Does it remove the stderr massage from the log?

**Comment by lgeiger** [comment]
> Does it remove the stderr message from the log?

No, I get the same message from IJavascript.

Note that I only see the logs from IJavascript because I enabled them via `process.env.DEBUG='*'` and IJavascript logs them to `stderr`.

**Comment by BenRussert** [comment]
I get the same startup issue w/ IJavascript. I've noticed the issue does not occur when starting a kernel using the command palette 'hydrogen:select-kernel' ( before running any code ).

If I try to run/runAll and automatically start the kernel the issue occurs probably 2/3 of the time

**Comment by n-riesco** [comment]
@BenRussert Do you get any error message? If so, could you post here.

Also could you post the IJavascript version and the kernel spec you're using?

To list the kernel specs, run `jupyter kernelspec list --json`.

**Comment by BenRussert** [comment]
[Here is a screenshare of the issue](https://youtu.be/sTzsLxzCt5U?t=25) at the end I show the `hydrogen:select-kernel` command which allows everything to work as expected.

Below is my kernelspec:

```
{
  "kernelspecs": {
    "python3": {
      "resource_dir": "/home/ben/anaconda3/envs/benr/lib/python3.5/site-packages/ipykernel/resources",
      "spec": {
        "display_name": "Python 3",
        "env": {},
        "language": "python",
        "argv": [
          "/home/ben/anaconda3/envs/benr/bin/python",
          "-m",
          "ipykernel",
          "-f",
          "{connection_file}"
        ]
      }
    },
    "javascript": {
      "resource_dir": "/home/ben/.local/share/jupyter/kernels/javascript",
      "spec": {
        "display_name": "Javascript (Node.js)",
        "env": {},
        "language": "javascript",
        "argv": [
          "ijskernel",
          "{connection_file}",
          "--protocol=5.0"
        ]
      }
    },
    "julia-0.5": {
      "resource_dir": "/home/ben/.local/share/jupyter/kernels/julia-0.5",
      "spec": {
        "display_name": "Julia 0.5.0",
        "env": {},
        "language": "julia",
        "argv": [
          "/opt/julia5/bin/julia",
          "-i",
          "--startup-file=yes",
          "--color=yes",
          "/home/ben/.julia/v0.5/IJulia/src/kernel.jl",
          "{connection_file}"
        ]
      }
    },
    "python2": {
      "resource_dir": "/home/ben/.local/share/jupyter/kernels/python2",
      "spec": {
        "display_name": "Python 2",
        "env": {},
        "language": "python",
        "argv": [
          "/home/ben/anaconda3/envs/ipykernel_py2/bin/python",
          "-m",
          "ipykernel",
          "-f",
          "{connection_file}"
        ]
      }
    }
  }
}
```

**Comment by n-riesco** [comment]
@BenRussert Thanks for the log, it is really useful. The kernel spec is OK, but the atom log shows something unusual.

@lgeiger Have a look at Ben's video (at 1:00). The unusual thing I refer above is that:
- Hydrogen invokes `Kernel.execute` before the shellSocket is logged as connected, which should be OK (that's what we use zmq is for), but it seems it isn't.
- Hydrogen gets IO messages before the ioSocket is logged as connected (weird, but possible depending on the order of things being scheduled in the event loop).

At the time I was debugging #88, I tried to send `kernel_info_request` and wait for a reply before invoking `Kernel.execute`. This didn't fully resolve the problem (I could still see the occasional failure).

`setTimeout` (as you discuss above) could also work, but I'm afraid it'd be unreliable and fail sometimes.

The way I currently understand the Jupyter messing protocol is that:
- synchronisation with the IOSocket is guaranteed by waiting for a `execution_state: 'starting'` (unfortunately, many kernels don't implement this message; e.g. IRKernel).
- synchronisation with the ShellSocket could be achieved by sending `kernel_info_request` (but again, not every kernel implements it; e.g. IJulia).

I like the idea of waiting until zeromq reports that the sockets are connected. BTW, since zeromq uses zmq v4, I don't think we need a try block [here](https://github.com/nteract/hydrogen/blob/master/lib/zmq-kernel.js#L104-L116) any longer.

**Comment by lgeiger** [comment]
@BenRussert Thanks for the video. I can reproduce the same behavior.

@n-riesco Thanks for taking a closer look!
> Hydrogen invokes Kernel.execute before the shellSocket is logged as connected, which should be OK (that's what we use zmq is for), but it seems it isn't.

Indeed that should be OK, a successful kernel start shows the same behavior. See the [logs of a succesful start](https://cloud.githubusercontent.com/assets/13285808/22178371/a1ab795a-e034-11e6-9a2c-cf0749c6a87e.png) from above.

> Hydrogen gets IO messages before the ioSocket is logged as connected (weird, but possible depending on the order of things being scheduled in the event loop).

I would agree with "weird but possible" 😄 

> setTimeout (as you discuss above) could also work, but I'm afraid it'd be unreliable and fail sometimes.

Yeah, I don't like this solution either since it seams to only happen for IJavascript (at least on my machine).

> I like the idea of waiting until zeromq reports that the sockets are connected.

Definitely a thing we could try.

> BTW, since zeromq uses zmq v4, I don't think we need a try block here any longer.

Thanks for pointing this out.

**Comment by n-riesco** [comment]
@BenRussert @lgeiger Just an idea:
- does the problem disappear if you install the IJavascript kernel with `ijs --ijs-spec-path=full`?
- does the kernel spec point to the same file you get when you run `require("child_process").execSync("which ijskernel").toString();` in atom?

**Comment by lgeiger** [comment]
Thanks for the suggestions!

> does the problem disappear if you install the IJavascript kernel with `ijs --ijs-spec-path=full`?

Unfortunately not 😕 

> does the kernel spec point to the same file you get when you run require("child_process").execSync("which ijskernel").toString(); in atom?

The kernelspec points to:
```bash
"javascript": {
  "resource_dir": "/Users/lukasgeiger/Library/Jupyter/kernels/javascript", 
  "spec": {
    "language": "javascript", 
    "display_name": "Javascript (Node.js)", 
    "env": {}, 
    "argv": [
      "/usr/local/Cellar/node/7.7.3/bin/node", 
      "/usr/local/lib/node_modules/ijavascript/lib/kernel.js", 
      "{connection_file}", 
      "--protocol=5.0"
   ]
}
```

`which ijskernel` returns `/usr/local/bin/ijskernel` which is a symlink to the file mentioned in the kernelspec.

**Comment by n-riesco** [comment]
@lgeiger It was a long shot, anyway. Thanks for testing.

What surprises me the most in this case is that unlike #88, IJavascript is able to send back some messages on the IOSocket, but not the result...

Does it happen regardless of the piece of code you execute?

Could you run the IJavascript tests (`npm t`) on that machine?

**Comment by BenRussert** [comment]
@n-riesco I've had similar results with a number of code samples. If you want me to run a particular snippet fire away and I'll try it out. I will run `npm t` after work in a few hours.

Just a shot in the dark: I notice @lgeiger is running node v7.x and I am on v6.x, in one of your comments above you mention v4.x, any potential issues here?

**Comment by lgeiger** [comment]
> Could you run the IJavascript tests (npm t) on that machine?

Thanks for the hint! The tests totally fail on my machine. I don't know why I didn't tried this earlier 😄 

Just realized that IJavascript doesn't use Travis, I try to reproduce it there and make a PR.

**Comment by lgeiger** [comment]
The test output on my machine is either https://travis-ci.org/lgeiger/ijavascript/jobs/213530248
or https://travis-ci.org/lgeiger/ijavascript/jobs/213530249.

The first output probably shows the problem described in this issue.

**Comment by BenRussert** [comment]
I get similar results as @lgeiger above when testing.

**Comment by lgeiger** [comment]
Fixed in #678 🚀
